pages/search/index.js

```diff
@@ -43,79 +45,88 @@ const flex = new FlexSearch({
   }
 });
 
-export default function Search({pageData, menuData}) {
-  const [filter, setFilter] = useState('');
-  const [results, setResults] = useState(null);
-  const [flexindex, setFlexIndex] = useState(flex);
-  const [sermonflexindex, setSermonFlexIndex] = useState(sermonflex);
+async function retrieveIndexes(url) {
+  const data = await ky(url, {prefixUrl: 'http://localhost:3000'}).json();
 
-  async function handleChange(event) {
-    const filter = event.target.value;
-    const results = await flexindex
-      .search({
+  await flex.import(data.mainIndexExport);
+  await sermonflex.import(data.sermonIndexExport);
+
+  return {
+    async query(filter) {
+      const searchResults = await flex.search({
         query: filter,
         limit: 10,
         threshold: 5,
         depth: 3,
         field: ['title', 'searchbody']
-      })
-      .concat(
-        sermonflexindex.search({
-          query: filter,
-          limit: 10,
-          threshold: 5,
-          depth: 3,
-          field: ['title', 'preacher', 'book', 'series']
-        })
-      );
-    setResults(results);
-    setFilter(filter);
+      });
+
+      // const sermonResults = await sermonflex.search({
+      //   query: filter,
+      //   limit: 10,
+      //   threshold: 5,
+      //   depth: 3,
+      //   field: ['title', 'preacher', 'book', 'series']
+      // });
+
+      const results = searchResults; // .concat(sermonResults);
+
+      return results;
+    }
   };
+}
 
-  async function retrieveIndexes() {
-    const data = await fetch('http://localhost:3000/api/searchindex');
-    const parsed = await data.json();
-    await flex.import(parsed.mainIndexExport);
-    await sermonflex.import(parsed.sermonIndexExport);
-  }
+function useSearchIndex(filter) {
+  const {data: index} = useQuery('/api/searchindex', retrieveIndexes);
+  const {data: results} = useQuery(filter, filter => {
+    if (index) {
+      return index.query(filter);
+    }
+  });
```

Comment:
`useQuery()` generally takes two arguments

The first one is the query key, which it uses to cache queries. So if you search the same term twice, it will only run the query once, but return the same result each time

The second argument takes any function that returns promise (fetch, flexsearch, etc..), aka an async function. It caches the result of this function using the query key

The great thing is `useQuery('/api/searchindex', retrieveIndexes)` only fetches the index once, because the query key `'/api/searchindex'` never changes 🎉